### ISSUE TITLE
Cannot swap package with itself

### ISSUE THREAD

**Issue by evan-goode** [rpm-software-management/dnf5#1368]
This may not be a valid issue, but DNF 5 differs from DNF 4 here.

If you try to swap a package with itself, DNF 4 will print a "conflicting requests" warning, but will proceed to reinstall the package:

```
$ sudo dnf4 swap plocate plocate
Package plocate-1.1.22-1.fc39.x86_64 is already installed.
Dependencies resolved.

 Problem: cannot install the best candidate for the job
  - conflicting requests
==========================================================================================================================================
 Package                        Architecture                  Version                                Repository                      Size
==========================================================================================================================================
Reinstalling:
 plocate                        x86_64                        1.1.22-1.fc39                          updates                        173 k

Transaction Summary
==========================================================================================================================================
```

Or, similarly, swapping a package with something it obsoletes (`plocate` obsoletes `mlocate`):

```
$ sudo dnf4 swap plocate mlocate
Package plocate-1.1.22-1.fc39.x86_64 is already installed.
Dependencies resolved.

 Problem: cannot install the best candidate for the job
  - conflicting requests
======================================================================================================================================
 Package                       Architecture                 Version                               Repository                     Size
======================================================================================================================================
Reinstalling:
 plocate                       x86_64                       1.1.22-1.fc39                         updates                       173 k

Transaction Summary
======================================================================================================================================
```

In contrast, DNF 5 will print a similar error and just quit:

```
$ sudo dnf5 swap plocate plocate
Failed to resolve the transaction:
Package "plocate-1.1.22-1.fc39.x86_64" is already installed.
Problem: cannot install the best candidate for the job
  - conflicting requests
```

```
sudo dnf5 swap plocate mlocate
Failed to resolve the transaction:
Problem: cannot install the best candidate for the job
  - conflicting requests
```

IMO the behavior of DNF 4 is desirable.

**Comment by evan-goode** [comment]
Described by baryluk during the F40 DNF 5 test event: https://testdays.fedoraproject.org/events/180

**Comment by j-mracek** [comment]
I am really sorry, but I cannot reproduce the issue on Fedora 38. 
```
rpm -q libsolv dnf5
libsolv-0.7.28-20240402005049.26.g682a0001.fc38.x86_64
dnf5-5.1.15-20240402005101.39.gc1565a07.fc38.x86_64
```


```
$ dnf5 swap plocate plocate
Updating and loading repositories:
Repositories loaded.
Package "plocate-1.1.22-1.fc38.x86_64" is already installed.
Problem: cannot install the best candidate for the job
  - conflicting requests

Package                        Arch      Version                        Repository          Size
Reinstalling:                                                                                   
 plocate                       x86_64    1.1.22-1.fc38                  updates        530.4 KiB
   replacing plocate           x86_64    1.1.22-1.fc38                  updates        530.4 KiB

Transaction Summary:
 Reinstalling:      1 packages
 Replacing:         1 packages

Total size of inbound packages is 172 KiB. Need to download 172 KiB.
After this operation 0 B will be used (install 530 KiB, remove 530 KiB).
Is this ok [y/N]: 
```

When I excluded available packages then I've got reported behavior, but it is correct.
```
dnf5 swap plocate plocate --setopt=*.excludepkgs=plocate
Updating and loading repositories:
Repositories loaded.
Failed to resolve the transaction:
Package "plocate-1.1.22-1.fc38.x86_64" is already installed.
Problem: conflicting requests
```

Without additional information I am proposing to close the issue. Debugsolv data might be useful.

**Comment by j-mracek** [comment]
Interesting - the issue is reproducible with DNF and DNF5, but the reproducer requires `--best` for both.

**Comment by j-mracek** [comment]
The problem is that the best candidate for `job install oneof` is installed package `plocate-1.1.22-1.fc39.x86_64@@System` but this object is supposed to be removed by `job erase pkg plocate-1.1.22-1.fc39.x86_64@@System [cleandeps]`.

```
job install oneof plocate-1.1.22-1.fc39.x86_64@@System plocate-1.1.17-3.fc39.x86_64@fedora-modular plocate-1.1.17-3.fc39.x86_64@updates-modular plocate-1.1.22-1.fc39.x86_64@updates plocate-1.1.17-3.fc39.x86_64@fedora [forcebest,setevr,setarch]
job erase pkg plocate-1.1.22-1.fc39.x86_64@@System [cleandeps]
```

The problem can be resolved by removal of `plocate-1.1.22-1.fc39.x86_64@@System` from `job install oneof` but it will create another problem when package that should go to the system will be only available.

I believe that the issue is already resolved by suggestion to try `--no-best`:
```
Failed to resolve the transaction:
Package "plocate-1.1.22-1.fc39.x86_64" is already installed.
Problem: cannot install the best candidate for the job
  - conflicting requests
Debug data written to "/home/jmracek/debugdata"
You can try to add to command line:
  --no-best to not limit the transaction to the best candidates
```

I am closing the issue as resolved by a suggestion in the output.
